File: 1201-1300/1231-Divide Chocolate/1231-Divide Chocolate.cpp
class Solution {
public:
    int maximizeSweetness(vector<int>& sweetness, int K) {
        int low = 1, high = accumulate(sweetness.begin(), sweetness.end(), 0) / (K + 1);
        while (low < high) {
            int mid = low + (high - low + 1) / 2;
            int count = 0, curr = 0;
            for (int s : sweetness) {
                curr += s;
                if (curr >= mid) {
                    if (++count >= K + 1) {
                        break;
                    }
                    curr = 0;
                }
            }
            
            if (count >= K + 1) {
                low = mid;
            }
            else {
                high = mid - 1;
            }
        }
        
        return low;
    }
};

File: 1201-1300/1231-Divide Chocolate/1231-Divide Chocolate.py
class Solution:
    def maximizeSweetness(self, sweetness: List[int], K: int) -> int:
        low, high = 1, sum(sweetness) // (K + 1)
        while low < high:
            mid = (low + high + 1) // 2
            count = curr = 0
            for s in sweetness:
                curr += s
                if curr >= mid:
                    count += 1
                    if count >= K + 1:
                        break
                    curr = 0
                    
            if count >= K + 1:
                low = mid
            else:
                high = mid - 1

        return low
